pkg/recovery/recover.go

```diff
@@ -0,0 +1,332 @@
+		return nil, err
+	} else {
+		as = append(as, objAs...)
+	}
+	if objAs, err := serializeListObjToYAML(&cp.services); err != nil {
+		return nil, err
+	} else {
+		as = append(as, objAs...)
+	}
+	if objAs, err := serializeListObjToYAML(&cp.serviceAccounts); err != nil {
+		return nil, err
+	} else {
+		as = append(as, objAs...)
+	}
+	return as, nil
+}
+
+// renderBootstrap returns assets for a bootstrap control plane that can be used with `bootkube
+// start` to re-bootstrap a control plane. These assets are derived from the self-hosted control
+// plane that was recovered by the backend, but modified for direct injection into a kubelet.
+func (cp *controlPlane) renderBootstrap(kubeConfigPath string) (asset.Assets, error) {
+	// Extract pod specs from daemonsets and deployments.
+	var pods []*v1.Pod
+	for _, ds := range cp.daemonSets.Items {
+		if componentName := ds.Labels["component"]; bootstrapComponents[componentName] {
+			pod := &v1.Pod{Spec: ds.Spec.Template.Spec}
+			if err := setBootstrapPodMetadata(pod, ds.ObjectMeta); err != nil {
+				return nil, err
+			}
+			pods = append(pods, pod)
+		}
+	}
+	for _, ds := range cp.deployments.Items {
+		if componentName := ds.Labels["component"]; bootstrapComponents[componentName] {
+			pod := &v1.Pod{Spec: ds.Spec.Template.Spec}
+			if err := setBootstrapPodMetadata(pod, ds.ObjectMeta); err != nil {
+				return nil, err
+			}
+			pods = append(pods, pod)
+		}
+	}
+
+	// Fix up the pods.
+	var as asset.Assets
+	requiredSecrets := make(map[string]bool)
+	for _, pod := range pods {
+		// Change secret volumes to point to file mounts.
+		for i := range pod.Spec.Volumes {
+			vol := &pod.Spec.Volumes[i]
+			if vol.Secret != nil {
+				requiredSecrets[vol.Secret.SecretName] = true
+				secretPath := path.Join(asset.BootstrapSecretsDir, vol.Secret.SecretName)
+				vol.HostPath = &v1.HostPathVolumeSource{Path: secretPath}
+				vol.Secret = nil
+			}
+		}
+
+		// Make sure the kubeconfig is in the commandline.
+		for i, _ := range pod.Spec.Containers {
+			cn := &pod.Spec.Containers[i]
+			// Assumes the bootkube naming convention is used. Could also just make sure the image uses hyperkube.
+			if kubeConfigComponents[cn.Name] {
+				cn.Command = append(cn.Command, "--kubeconfig=/kubeconfig/kubeconfig")
+				cn.VolumeMounts = append(cn.VolumeMounts, v1.VolumeMount{
+					MountPath: "/kubeconfig",
+					Name:      "kubeconfig",
+					ReadOnly:  true,
+				})
+			}
+		}
+
+		// Add a mount for the kubeconfig.
+		pod.Spec.Volumes = append(pod.Spec.Volumes, v1.Volume{
+			VolumeSource: v1.VolumeSource{HostPath: &v1.HostPathVolumeSource{Path: kubeletKubeConfigPath}},
+			Name:         "kubeconfig",
+		})
+
+		// Output the pod definition.
+		a, err := serializeYAML(path.Join(asset.AssetPathBootstrapManifests, pod.Name+".yaml"), pod)
+		if err != nil {
+			return nil, err
+		}
+		as = append(as, a)
+	}
+
+	// Output all the required secrets.
+	for _, secret := range cp.secrets.Items {
```

Comment:
Todo is fine - but we should have a follow-up soon supporting configMaps. It's possible that the control-plane requires them (then the recovery wouldn't work).